lib/immutable-struct.rb

```diff
@@ -67,9 +85,37 @@ def self.new(*attributes,&block)
     end
     klass.class_exec(&block) unless block.nil?
     imethods = klass.instance_methods(include_super=false)
-    klass.class_exec(imethods) do |imethods|
+    klass.class_exec(imethods, attributes) do |imethods, attributes|
+
+      derived_methods = imethods - attributes - [:to_h]  #to_h is never a derived method, despite its 0-arity
+
+      ##
+      # :method: attributes_to_h
+      # Return a +Hash+ with only those attributes defined in the class constructor.
+      define_method(:attributes_to_h) do
+        attributes.each_with_object({}) do |attr, hash|
+          ivar = attr.to_s.prepend(?@).to_sym
+          hash[attr] = self.instance_variable_get(ivar)
+        end
+      end
+
+      ##
+      # :method: derived_to_h
+      # Return a +Hash+ of the results all zero-arity methods that are not attributes
+      # defined in the class constructor 
+      define_method(:derived_to_h) do
+        derived_methods.each_with_object({}) do |derived_method, hash|
+          hash[derived_method] = self.send(derived_method) if 0 == self.method(derived_method).arity
+        end
+      end
+
+      alias_method :__to_h__, :to_h if method_defined?(:to_h)
+      ##
+      # :method: to_h
+      # Return a +Hash+ with the result of merging +attributes_to_h+ and +derived_to_h+.
+      # This method can also be overriden so that, for example, only +attributes_to_h+ is used.
       define_method(:to_h) do
-        imethods.inject({}){ |hash, method| hash.merge(method.to_sym => self.send(method)) }
+        self.respond_to?(:__to_h__) ? __to_h__ : attributes_to_h.merge(derived_to_h)
```

Comment:
OK. This is just in case the user decides to override the method. There's already a Spec for this, using:

```ruby
let(:klass_with_overriden_to_h_custom) {
      ImmutableStruct.new(:flappy) do
        def lawsuit
          'pending'
        end
        def to_h
          {:custom => 'hash'}
        end
      end
    }

```